src/scheduler.ml

```diff
@@ -190,18 +195,109 @@ let go ?(log=Log.no_log) ?(config=Config.default)
     { log
     ; gen_status_line
     ; original_cwd = cwd
-    ; display      = config.display
+    ; display      = config.Config.display
     ; concurrency  = (match config.concurrency with Auto -> 1 | Fixed n -> n)
     ; status_line  = ""
     ; waiting_for_available_job = Queue.create ()
     }
   in
   Errors.printer := print t;
+  t
+
+let run t fiber =
   let fiber =
     Fiber.Var.set t_var t
-      (Fiber.with_error_handler (fun () -> fiber) ~on_error:Report_error.report)
+      (Fiber.with_error_handler fiber ~on_error:Report_error.report)
   in
   Fiber.run
     (Fiber.fork_and_join_unit
        (fun () -> go_rec t)
        (fun () -> fiber))
+
+let go ?log ?config ?gen_status_line fiber =
+  let t = prepare ?log ?config ?gen_status_line () in
+  run t (fun () -> fiber)
+
+(** Fiber loop looks like this (if cache_init is true):
+              /------------------\
+              v                  |
+    init --> once --> finally  --/
+
+    The result of [~init] gets passed in every call to [~once] and [~finally].
+    If cache_init is false, every iteration reexecutes init instead of
+    saving it.
+
+    [~watch] should return after the first change to any of the project files.
+*)
+let poll ?log ?config ?(cache_init=true) ~init ~once ~finally ~watch () =
+  let t = prepare ?log ?config () in
+  let saved_setup = ref None in
+  let setup () =
+    match !saved_setup with
+    | Some c -> c
+    | None ->
+      Format.eprintf "Internal error: setup failed.\n";
+      exit 1
+  in
+  let init =
```

Comment:
This should be `init ()`. In general we assume that a given `Fiber.t` value is executed only once